Android TV : added Google cast support  that brings media title,album,artist,artwork,position,duration and seeking support

albaintor opened the pull request:
Thanks to this home assistant library https://github.com/home-assistant/core/tree/dev/homeassistant/components/cast this is now possible to retrieve additional media info :

- Media title
- Artist
- Album
- Position
- Duration
- Artwork

Also I added a support for a seeking method to set media position.

I integrated the library as a new requirement and trigger it in the connect method.
I was able to test it successfully : now you have the mentioned info displayed on the screen (media player widget).

Not all the apps provide media info:  tested successfully with disney+ and netflix.
The artwork is missing in all cases even though it is retrieved and published

I didn't dig all the possibilities yet. Also at last I think that this library should be migrated to asyncio.

albaintor commented:
Hi Markus,

I have made further tests and I had to rewrite the library : it was using threads and didn't worked correctly with the event loop of the integration. I migrated the code to asyncio and submitted a PR to pychromecast https://github.com/home-assistant-libs/pychromecast/pull/1048

In the meantime, I imported the modified chromecast library into the integration. It may be removed later and switched when the PR will be merged.

Comparing to the last commit :

- The media state is correctly updated (playing, paused, idle...)
- The media title is now taken from 2 origins : the media title from chromecast if not empty, or the current application

The best app to test it with is Disney+ : it provides media title, state, position...
Unfortunately, some other apps like Netflix doesn't provide any information

albaintor commented:
EDIT : fixed formatting
I also added a checkbox in the setup flow to enable chromecast features (enabled by default)

zehnm commented:
This is a great addition 👍 
Please let me know when this is ready to review, since you have been pushing more commits.

albaintor commented:
Yes this time this is it !
I just added volume support (up/down/mute and set level)
Seeking works well too (disney+ or appletv+)

Available info :
- Disney+ : media position, duration and title
- AppleTV+ : media position & duration
- Netflix : nothing
- Prime Video : nothing

It is working well but some more testing is needed : when wifi disconnect, after a long period of inactivity... I am not sure that the library notifies or handles automatic reconnection

zehnm commented on intg-androidtv/tv.py:
```diff
@@ -432,6 +488,24 @@ def _handle_invalid_auth() -> None:
+                self._chromecast.register_connection_listener(self)
+            try:
+                if not self._chromecast.connection_client.connected:
+                    await self._chromecast.connect(timeout=5)
```
Replace magic number with the already available `CONNECTION_TIMEOUT` constant for the regular Android TV connection. Then it's easier to change and eventually make it a configuration option.

albaintor replied:
Done

zehnm commented on intg-androidtv/profiles.py:
The default device profile may not contain the optional Google Cast features.
They have to be dynamically added if the feature is active (`AtvDevice.use_chromecast`). Otherwise commands are exposed to the Remote that are not available, but the user can still use them. This either leads to errors, or non-working actions.

The `match` method below needs to be enhanced:
- add `use_chromecast` parameter
- if True: add the new features to the matched profile

albaintor replied:
Done in both PRs

zehnm commented on config/profiles/Google_Chromecast.json:
See `profiles.py`: features may only be added if the Cast feature is enabled by the user.

zehnm commented on config/profiles/NVIDIA_SHIELD.json:
See `profiles.py`: features may only be added if the Cast feature is enabled by the user.

albaintor replied:
Done in both PRs

zehnm commented on config/profiles/default.json:
See `profiles.py`: features may only be added if the Cast feature is enabled by the user.

zehnm commented on intg-androidtv/config.py:
```diff
@@ -250,6 +252,9 @@ def migration_required(self) -> bool:
 
     async def migrate(self) -> bool:
         """Migrate configuration if required."""
+        # pylint: disable=C0415,R0401
+        from tv import AndroidTv
```
To avoid this hack and cyclic import - if we keep the `AtvDevice` usage in `AndroidTv` -  the `migrate` method should be refactored as an external function in a separate file.

We can keep it for the moment and focus on the main functionality, even though I really don't like cyclic imports. It's usually only a matter of time until the "design" doesn't work anymore and requires more ugly hacks.

zehnm commented on intg-androidtv/setup_flow.py:
```diff
@@ -44,6 +44,8 @@ class SetupSteps(IntEnum):
 _cfg_add_device: bool = False
 _discovered_android_tvs: list[dict[str, str]] = []
 _pairing_android_tv: tv.AndroidTv | None = None
+_use_chromecast: bool = True
```
Please change default to `False`:
Until we have this feature better tested and are sure that reconnection works, I'd like to release it as a preview feature.

zehnm commented on intg-androidtv/setup_flow.py:
```diff
@@ -335,7 +337,15 @@ async def _handle_discovery(msg: UserDataResponse) -> RequestUserInput | SetupEr
+                    "en": "Enable Chromecast features",
+                    "fr": "Activer les fonctionnalités de Chromecast",
+                },
+                "field": {"checkbox": {"value": True}},
```
See above about default.

zehnm commented on intg-androidtv/tv.py:
```diff
@@ -482,38 +561,53 @@ def _is_on_updated(self, is_on: bool) -> None:
+            # Chromecast service is not accessible when the device is in standby
+            try:
+                if self._chromecast and not self._chromecast.connection_client.connected:
+                    asyncio.create_task(self._chromecast.connect(timeout=5))
```
Magic number: see above

zehnm reviewed (changes requested):
This is an initial code review only. I'll do a functional test next, focusing an Google Cast reconnection behaviour.

Some issues already need to be addressed. In short:

- new profile features must be added dynamically if the user enabled Google Cast.
- default setting of Google Cast should be off.
  I consider this a testing feature at the moment. Once throughly tested and working, I'm happy to make it the default.
- replace hardcoded timeout with the available timeout constant.

If your pychromecast PR is not accepted, we have to come up with a solution.
Keeping a forked version inside this integration will likely create future maintenance and support work, which we can't handle. The pychromecast project seems to be quite active.
Otherwise I highly prefer using the original version to easily keep up with future releases and bug fixes.
What were your issues with the library using threads? Maybe there's still a way to integrate it.

Please add the missing pychromecast license information, so it's clear it's not our code and we properly attribute it. I suggest adding the LICENSE and README.rst files to the pychromecast directory.

zehnm commented on intg-androidtv/tv.py:
```diff
@@ -637,3 +731,136 @@ async def _switch_input(self, source: str) -> ucapi.StatusCodes:
+            return ucapi.StatusCodes.NOT_IMPLEMENTED
+        try:
+            self._muted = not self._muted
+            self._chromecast.set_volume_muted(self._muted)
```
missing await

albaintor commented on intg-androidtv/tv.py:
```diff
@@ -128,17 +164,16 @@ async def wrapper(self: _AndroidTvT, *args: _P.args, **kwargs: _P.kwargs) -> uca
     return wrapper
 
 
-class AndroidTv:
+# pylint: disable=too-many-public-methods
+class AndroidTv(CastStatusListener, MediaStatusListener, ConnectionStatusListener):
     """Representing an Android TV device."""
 
     # pylint: disable=R0917
     def __init__(
         self,
         certfile: str,
         keyfile: str,
-        host: str,
-        name: str,
-        identifier: str | None = None,
+        device_config: AtvDevice,
```
Yes this is something I always do now - not cyclic imports  : using the device_config structure instead of passing parameters one by one. This makes things easier when you create new parameters. But indeed (only) in this integration there is the migration logic that requires to access to the device instance class. To be fixed (I didn't figure out an easy way)

albaintor commented on .idea/integration-androidtv.iml:
Ok reverted

albaintor commented:
Hi Markus,
I have created a second PR https://github.com/unfoldedcircle/integration-androidtv/pull/57 using the current (external) library

Both PR are working, so one with the copied and modified library with asyncio, and the other one with the current release.
I have applied the same modifications to both : dynamic features injected based on `use_chromecast` parameter (now disabled by default)

You will be able to test if it is working well on your side.
On my side (but I don't remember exactly where/when), after a while I had no more updates. I suspected wrong sync between the main thread and the thread of the socket.
Still, these issues could be related to bugs I fixed in the meantime on the driver side, so I am not sure anymore.

Anyway, this pychromecast PR will be good and hopefully make it through as it is more efficient (asyncio in a unique thread vs infinite loop on socket receive in a second thread)

zehnm commented:
With the latest update I no longer get media title information anymore. In the logs the following warnings are continuously repeated:
```
WARNING:asyncio:SSL connection is closed
WARNING:asyncio:SSL connection is closed
WARNING:asyncio:SSL connection is closed
DEBUG:tv:[ShieldTV Wohnzimmer] Received Chromecast connection status : ConnectionStatus(status='LOST', address=NetworkAddress(address='xxx', port=8009), service=None)
```

I switch over to the new PR and test that one.
At the same time I see the correct media information in the Google Cast Home Assistant integration.
Restarting the integration driver fixed it. It's most likely a reconnection issue where multiple connections are trying to reconnect instead of just one.

albaintor commented:
Ok I will look at it tonight back from work

albaintor commented:
Hi again, I could test it and it doesn't work either.
I applied this modification and it works again. I'll do more tests

albaintor commented:
I cancel this PR as the other PR with legacy library works
